quantization: autogenerate quantization backend configs for documentation

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#75126 quantization: autogenerate quantization backend configs for documentation**

Summary:

Quantization has a high volume of configurations of how to quantize an
op for a reference model representation which is useful for a lowering
step for a backend.  An example of this is

```
 {'dtype_configs': [{'input_dtype': torch.quint8,
										 'output_dtype': torch.quint8}],
	'observation_type': <ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT: 0>,
	'pattern': <class 'torch.nn.modules.conv.ConvTranspose1d'>},
```

These configs are checked into master, and they are created with Python functions.
Therefore, there is no easy way for the user to see what the configs actually
are without running some Python code.

This PR is one approach to document these configs. Here is what this is doing:
1. during documentation build, write a text file of the configs
2. render that text file on a quantization page, with some additional context

In the future, this could be extended to autogenerate better looking tables
such as: op support per backend and dtype, op support per valid quantization settings per backend,
etc.

Test plan:

```
cd docs
make html
cd html
python -m http.server 8000
// render http://[::]:8000/quantization-backend-configuration.html
// it renders correctly
```

Differential Revision: [D35365461](https://our.internmc.facebook.com/intern/diff/D35365461)

vkuzo committed:
[rfc] quantization: autogenerate quantization backend configs for documentation

Summary:

Quantization has a high volume of configurations of how to quantize an
op for a reference model representation which is useful for a lowering
step for a backend.  An example of this is

```
 {'dtype_configs': [{'input_dtype': torch.quint8,
										 'output_dtype': torch.quint8}],
	'observation_type': <ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT: 0>,
	'pattern': <class 'torch.nn.modules.conv.ConvTranspose1d'>},
```

These configs are checked into master, and they are created with Python functions.
Therefore, there is no easy way for the user to see what the configs actually
are without running some Python code.

This PR is one approach to document these configs. Here is what this is doing:
1. during documentation build, write a text file of the configs
2. render that text file on a quantization page, with some additional context

In the future, this could be extended to autogenerate better looking tables
such as: op support per backend and dtype, op support per valid quantization settings per backend,
etc.

Currently this is an RFC to ensure there aren't any security concerns with
rendering a text file. If there aren't, this could move to review by
quantization team.

Note: the content being rendered is a placeholder, for now. We need a better
looking print-out, with more context, before we can ship something like
this to users.

Test plan:

```
cd docs
make html
cd html
python -m http.server 8000
// render http://[::]:8000/quantization-backend-configuration.html
// it renders correctly
```

[ghstack-poisoned]

andrewor14 commented on docs/source/quantization-backend-configuration.rst:
```diff
@@ -0,0 +1,26 @@
+
+  from torch.ao.quantization.fx.backend_config import get_native_backend_config_dict
+  from pprint import pprint
+  pprint(get_native_backend_config_dict())
```
I think this would look better:
```
import json
print(json.dumps(get_native_backend_config_dict(), indent=4))
```

vkuzo replied:
Something like this would work:

```
from torch.utils._pytree import tree_map
config = get_native_backend_config_dict()
# convert Python types to strings, so json module can understand them
config_str = tree_map(str, config)
f.write(json.dumps(config_str, indent=2))
```

However, the results would look like:

```
{
  "pattern": "<function leaky_relu at 0x7fd4083abb80>",
  "observation_type": "ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT",
  "dtype_configs": [
    {
      "input_dtype": "torch.quint8",
      "output_dtype": "torch.quint8"
    }
  ]
},
```

It's easier to read, but also everything is printed as strings, which may be confusing to the user since these objects are not actually strings.  I think a custom printing function may have to be written to both make it nice on the eyes and also not print incorrect types.

andrewor14 replied:
I see. Maybe we can do that in the future then

andrewor14 reviewed (approved):
By the way I just landed a bunch of new configs. Could you rebase and see what it looks like now?

jerryzh168 commented on docs/source/scripts/build_quantization_configs.py:
```diff
@@ -0,0 +1,23 @@
+# Create a directory for the images, if it doesn't exist
+QUANTIZATION_CONFIG_IMAGE_PATH = os.path.join(
+    os.path.realpath(os.path.join(__file__, "..")),
+    "quantization_configs"
```
maybe quantization_operator_config, or quantization_backend_config? since `quantization config` is already sued for QConfig..

vkuzo replied:
sure, changed

vkuzo committed:
Update on "quantization: autogenerate quantization backend configs for documentation"


Summary:

Quantization has a high volume of configurations of how to quantize an
op for a reference model representation which is useful for a lowering
step for a backend.  An example of this is

```
 {'dtype_configs': [{'input_dtype': torch.quint8,
										 'output_dtype': torch.quint8}],
	'observation_type': <ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT: 0>,
	'pattern': <class 'torch.nn.modules.conv.ConvTranspose1d'>},
```

These configs are checked into master, and they are created with Python functions.
Therefore, there is no easy way for the user to see what the configs actually
are without running some Python code.

This PR is one approach to document these configs. Here is what this is doing:
1. during documentation build, write a text file of the configs
2. render that text file on a quantization page, with some additional context

In the future, this could be extended to autogenerate better looking tables
such as: op support per backend and dtype, op support per valid quantization settings per backend,
etc.

Test plan:

```
cd docs
make html
cd html
python -m http.server 8000
// render http://[::]:8000/quantization-backend-configuration.html
// it renders correctly
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35365461).
